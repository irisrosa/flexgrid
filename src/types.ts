import React, { CSSProperties, ElementType, PropsWithChildren, ReactElement } from 'react';

import { DefaultTheme } from 'styled-components';

export type PropsWithTheme<T> = T & { theme: DefaultTheme };

export type BasicComponentProps = PropsWithChildren<{
  className?: string;
  style?: CSSProperties;
}>;

export type ImageType = {
  [key: string]: unknown;
  alt?: string;
  aspectRatio?: number;
  base64?: string;
  sizes?: string;
  src: string;
  srcSet?: string;
};

export type BackgroundProps = {
  renderCustomBackground?: () => ReactElement;
  backgroundImage?: ImageType;
};

export interface BlockData extends BackgroundProps {
  blockPadding?: false | string;
  Content: ElementType;
  size?: number;
  as?: keyof AllowedTags;
  offset?: number;
}

export type AllowedTags = Pick<
  HTMLElementTagNameMap,
  | 'address'
  | 'a'
  | 'article'
  | 'aside'
  | 'br'
  | 'colgroup'
  | 'col'
  | 'dl'
  | 'dt'
  | 'dd'
  | 'details'
  | 'div'
  | 'figure'
  | 'footer'
  | 'h1'
  | 'h2'
  | 'h3'
  | 'h4'
  | 'h5'
  | 'h6'
  | 'header'
  | 'img'
  | 'main'
  | 'nav'
  | 'ol'
  | 'li'
  | 'p'
  | 'picture'
  | 'section'
  | 'span'
  | 'summary'
>;

export type AllowedProps<T extends keyof AllowedTags> = React.ComponentPropsWithoutRef<T>;

export type GridOptions = {
  blockPadding?: false | string;
  /** Preferably use rem unit to set the gap  */
  gap?: false | string;
  columns?: number;
  maxWidth?: number;
  breakPoints?: {
    small: number;
    large: number;
  };
};

export type FlexGridProps = BackgroundProps &
  GridOptions & {
    blocks?: BlockData[];
    as?: keyof AllowedTags;
    extendBackground?: boolean;
    extendContent?: boolean;
  };

export type BlockSetProps = Omit<FlexGridProps, keyof GridOptions | 'blocks'>;

export type BlockProps = Partial<BlockData>;

export type PolymorphicRef<C extends React.ElementType> = React.ComponentPropsWithRef<C>['ref'];

export type PolymorphicComponentPropWithRef<
  C extends React.ElementType,
  Props = {}
> = React.PropsWithChildren<Props & { as?: C }> &
  Omit<React.ComponentPropsWithoutRef<C>, 'as'> & {
    ref?: PolymorphicRef<C>;
  };

export type PolymorphicComponent<P> = <E extends React.ElementType = 'div'>(
  props: PolymorphicComponentPropWithRef<E, P>
) => React.ReactElement;
