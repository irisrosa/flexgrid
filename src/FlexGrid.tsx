import React from 'react';
import { BlockSet } from '@components/layout/FlexGrid';
import { GridInterface } from './types';
import { withTheme } from '@src/hocs';
import { ThemeProvider } from 'styled-components';

export const FlexGridComponent: React.FC<GridInterface> = ({ data, theme }) => (
  <ThemeProvider theme={theme}>
    {Object.values(data).map(({ blocks, bleedBackground, ...props }, blockSetKey) => {
      return (
        <BlockSet
          data-testid="blockset"
          key={blockSetKey}
          blockSetStyle={bleedBackground ? blocks[0].style : 'default'}
          bleedBackground={bleedBackground}
          backgroundImage={blocks[0].useImageBackground && blocks[0].backgroundImage}
          blocks={blocks}
          {...props}
        />
      );
    })}
  </ThemeProvider>
);

export const FlexGrid: React.FC<GridInterface> = withTheme(FlexGridComponent);
