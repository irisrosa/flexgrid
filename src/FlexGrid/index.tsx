import React from 'react';

import { Blocks } from '@src/Blocks';

import { BlockSet } from '../BlockSet';
import { Grid } from '../types';

export const FlexGrid: React.FC<Grid> = ({ data }) => (
  <>
    {Object.values(data).map(({ blocks, ...props }, blockSetKey) => (
      <BlockSet key={blockSetKey} {...props}>
        <Blocks blocks={blocks} />
      </BlockSet>
    ))}
  </>
);
