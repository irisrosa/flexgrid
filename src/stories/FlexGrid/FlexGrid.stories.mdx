import { Meta, Story, Canvas, ArgsTable, Source } from '@storybook/addon-docs/blocks';
import dedent from 'ts-dedent';
import { FlexGrid, Block, BlockSet } from '@src';
import { data } from './data';

<Meta
  title="FlexGrid"
  component={FlexGrid}
  subcomponents={{
    Block,
    BlockSet,
  }}
  args={{
    data,
  }}
/>

export const Template = args => {
  return (
    <>
      <FlexGrid {...args} />
    </>
  );
};

# FlexGrid

The FlexGrid component renders groups of blocks (called `BlockSet`), positioning them accordingly to their options.

<Story name="Default">{Template.bind({})}</Story>

Example data that will generate a FlexGrid:

<Source dark language="javascript" code={dedent`${JSON.stringify(data, null, 2)}`} />
<Source
  dark
  language="jsx"
  code={dedent`
  <FlexGrid data={data} />
`}
/>

---

<ArgsTable of={FlexGrid} />

---

## Manual FlexGrid

It's also possible to build a FlexGrid combining `BlockSet` and `Block` components directly.

<Canvas>
  <Story name="Manual FlexGrid">
    <BlockSet>
      <Block>
        <p>
          Ut occaecat ut sunt veniam laborum quis velit pariatur ullamco esse duis velit aliquip. Do
          est aliqua et non. Nisi occaecat magna pariatur nulla aute mollit commodo dolor velit
          irure elit.
        </p>
        <button onClick={() => null}>Click me</button>
      </Block>
      <Block>
        <p>
          Ut occaecat ut sunt veniam laborum quis velit pariatur ullamco esse duis velit aliquip. Do
          est aliqua et non. Nisi occaecat magna pariatur nulla aute mollit commodo dolor velit
          irure elit.
        </p>
        <button onClick={() => null}>Click me</button>
      </Block>
      <Block>
        <p>
          Ut occaecat ut sunt veniam laborum quis velit pariatur ullamco esse duis velit aliquip. Do
          est aliqua et non. Nisi occaecat magna pariatur nulla aute mollit commodo dolor velit
          irure elit.
        </p>
        <button onClick={() => null}>Click me</butt>
      </Block>
      <Block>
        <p>
          Ut occaecat ut sunt veniam laborum quis velit pariatur ullamco esse duis velit aliquip. Do
          est aliqua et non. Nisi occaecat magna pariatur nulla aute mollit commodo dolor velit
          irure elit.
        </p>
        <button onClick={() => null}>Click me</button>
      </Block>
      <Block>
        <p>
          Ut occaecat ut sunt veniam laborum quis velit pariatur ullamco esse duis velit aliquip. Do
          est aliqua et non. Nisi occaecat magna pariatur nulla aute mollit commodo dolor velit
          irure elit.
        </p>
        <Button onClick={() => null}>Click me</Button>
      </Block>
    </BlockSet>
  </Story>
</Canvas>

## BlockSet

A BlockSet is a group of blocks with the same value of `bleed`.

<Canvas>
  <Story name="BlockSet">
    <BlockSet>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
    </BlockSet>
  </Story>
</Canvas>

A BlockSet with `bleedContent: true` will expand to the max width of its container or the viewport width, while the default `bleedContent: false` wraps its children in a container.

<Canvas>
  <Story name="BlockSet bleedContent">
    <BlockSet bleedContent>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
    </BlockSet>
  </Story>
</Canvas>

A BlockSet with `blockPadding: false` will remove the padding from the child `Block`.

<Canvas>
  <Story name="BlockSet no blockPadding">
    <BlockSet blockPadding={false}>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
    </BlockSet>
  </Story>
</Canvas>

### BlockSet Background

It's possible to set a background image covering the blockset.

<Canvas>
  <Story name="BlockSet with background image">
    <BlockSet
      bleedBackground
      backgroundImage={{
        src:
          'https://images.unsplash.com/photo-1600942956976-f5a7a7cd8ff7?crop=entropy&fm=jpg&h=900&q=80&w=1600&cs=tinysrgb&fit=crop',
        alt: 'Alt text',
      }}
    >
      <Block noBackground>
        <p>Block</p>
      </Block>
      <Block noBackground>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
    </BlockSet>
  </Story>
</Canvas>

<Canvas>
  <Story name="Blockset with custom background Component">
    <BlockSet
      blockSetStyle="dark"
      bleedBackground
      BgComp={({ children, ...props }) => (
        <>
          <img style={{ position: 'absolute' }} alt={props.alt} src={props.src} />
          {children}
        </>
      )}
      backgroundImage={{
        src:
          'https://images.unsplash.com/photo-1600942956976-f5a7a7cd8ff7?crop=entropy&fm=jpg&h=900&q=80&w=1600&cs=tinysrgb&fit=crop',
        alt: 'Alt text',
      }}
    >
      <Block noBackground>
        <p>Block</p>
      </Block>
      <Block noBackground>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
    </BlockSet>
  </Story>
</Canvas>

## Block

By default a `Block` wrapped by a `BlockSet` will adapt its width to be in the same row as its sibilings - up to 4 Blocks in the same row.

A block with `fullScreen: true` will expand to the limits of its `BlockSet`.

<Canvas>
  <Story name="Block fullWidth">
    <BlockSet>
      <Block fullWidth>
        <p>Block 100%</p>
      </Block>
      <Block>
        <p>Block 50%</p>
      </Block>
      <Block>
        <p>Block 50%</p>
      </Block>
    </BlockSet>
  </Story>
</Canvas>

### No Padding

Blocks can receive a prop `noPadding` that will remove its default padding around any children.

<Canvas>
  <Story name="Block no padding">
    <BlockSet>
      <Block noPadding>
        <p>Block no padding</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
    </BlockSet>
  </Story>
</Canvas>

### Block Background

It's possible to set a backgroung image to a block by passing the prop `backgroundImage`.

Also blocks can receive the a prop `noBackground` that will remove the default brand background color or image background.

<Canvas>
  <Story name="Block with background image">
    <BlockSet>
      <Block
        backgroundImage={{
          src:
            'https://images.unsplash.com/photo-1600942956976-f5a7a7cd8ff7?crop=entropy&fm=jpg&h=500&q=80&w=1600&cs=tinysrgb&fit=crop',
          alt: 'Alt Text',
        }}
      >
        <p>Block</p>
      </Block>
      <Block noBackground style={{ backgroundColor: '#CCC' }}>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
    </BlockSet>
  </Story>
</Canvas>

<Canvas>
  <Story name="Block with custom background Component">
    <BlockSet>
      <Block
        BgComp={({ children, ...props }) => (
          <>
            <img
              style={{ position: 'absolute', top: 0, left: 0 }}
              alt={props.alt}
              src={props.src}
            />
            {children}
          </>
        )}
        backgroundImage={{
          src:
            'https://images.unsplash.com/photo-1600942956976-f5a7a7cd8ff7?crop=entropy&fm=jpg&h=500&q=80&w=1600&cs=tinysrgb&fit=crop',
          alt: 'Alt Text',
        }}
      >
        <p>Block</p>
      </Block>
      <Block noBackground style={{ backgroundColor: '#CCC' }}>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
      <Block>
        <p>Block</p>
      </Block>
    </BlockSet>
  </Story>
</Canvas>

## Equalize heights

It's possible to equalize the heights of all blocks in a blockset or flexgrid by passing `equalizeHeights` prop as `true`.

<Canvas>
  <Story name="Equalize heights">
    <BlockSet>
      <Block style={{ background: 'grey' }}>
        <div style={{ background: 'lightgrey' }}>
          <p>Long</p>
          <p>text</p>
          <p>as content</p>
        </div>
      </Block>
      <Block style={{ background: 'grey' }}>
        <div style={{ background: 'lightgrey' }}>
          <p>Shorter</p>
          <p>content</p>
        </div>
      </Block>
      <Block style={{ background: 'grey' }}>
        <div style={{ flexGrow: 1, background: 'lightgrey' }}>
          <p>Small content equalized</p>
        </div>
      </Block>
    </BlockSet>
  </Story>
</Canvas>
