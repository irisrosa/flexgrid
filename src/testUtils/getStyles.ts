import { screen } from '@testing-library/react';

export const getStylesById = (dataTestId: string): Record<string, unknown> => {
  const element = screen.getByTestId(dataTestId);
  return getStyles(element);
};

export const getStyles = (element: HTMLElement): Record<string, unknown> => {
  return window.getComputedStyle(element)['_values'];
};
