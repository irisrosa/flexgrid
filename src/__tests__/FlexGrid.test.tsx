import React from 'react';
import { screen, render, within } from '@testing-library/react';
import { BlockSetData, GridInterface } from '@src/types';
import { FlexGrid } from '@src/FlexGrid';
import { getStyles } from '@src/testUtils';

const noOpComp = () => <React.Fragment />;

const noBleedBlockSet: BlockSetData = {
  bleedContent: false,
  bleedBackground: false,
  blockPadding: true,
  blocks: [
    { style: 'light', Component: noOpComp },
    { style: 'dark', Component: noOpComp },
    { style: 'light', Component: noOpComp },
  ],
};

const bleedbgBlockset: BlockSetData = {
  bleedContent: false,
  bleedBackground: true,
  blockPadding: true,
  blocks: [{ style: 'dark', Component: noOpComp }],
};

const fullBleedBlockSet: BlockSetData = {
  bleedContent: true,
  bleedBackground: true,
  blockPadding: false,
  blocks: [{ style: 'dark', Component: noOpComp }],
};

describe('FlexGrid', () => {
  test('defaults', () => {
    render(<FlexGrid data={[noBleedBlockSet, bleedbgBlockset, fullBleedBlockSet]} />);

    const allSets = screen.queryAllByTestId('blockset');
    const allBlocks = screen.queryAllByTestId('block');

    expect(screen.queryAllByTestId('blockset')).toHaveLength(3);
    expect(allSets.map(s => getStyles(s))).toMatchSnapshot();

    expect(screen.queryAllByTestId('block')).toHaveLength(5);
    expect(allBlocks.map(b => getStyles(b))).toMatchSnapshot();
  });

  test('background', () => {
    const blocksetWithBackground: BlockSetData = {
      bleedBackground: true,
      bleedContent: false,
      blockPadding: true,
      blocks: [
        {
          style: 'dark',
          noBackground: false,
          useImageBackground: true,
          backgroundImage: {
            src: 'http://somepath/image.png',
            alt: 'something',
          },
          Component: noOpComp,
        },
        {
          noBackground: true,
          Component: noOpComp,
        },
        {
          noBackground: true,
          Component: noOpComp,
        },
        {
          noBackground: true,
          Component: noOpComp,
        },
      ],
    };
    render(<FlexGrid data={[blocksetWithBackground]} />);

    const blockset = screen.getByTestId('blockset');
    const blockSetBg = within(blockset).queryAllByTestId('blockset-background-image');
    const blockBg = screen.queryAllByTestId('background-image');

    expect(screen.queryAllByTestId('blockset')).toHaveLength(1);
    expect(blockSetBg).toHaveLength(1);
    expect(blockBg).toHaveLength(0);
  });
});
