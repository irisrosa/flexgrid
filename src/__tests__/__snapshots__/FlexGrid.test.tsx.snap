// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexGrid defaults 1`] = `
Array [
  Object {
    "display": "block",
    "margin": "0px auto",
    "margin-bottom": "0px",
    "margin-left": "auto",
    "margin-right": "auto",
    "margin-top": "0px",
    "overflow": "hidden",
    "position": "relative",
    "visibility": "visible",
    "width": "100%",
  },
  Object {
    "display": "block",
    "overflow": "hidden",
    "position": "relative",
    "visibility": "visible",
  },
  Object {
    "display": "block",
    "overflow": "hidden",
    "position": "relative",
    "visibility": "visible",
  },
]
`;

exports[`FlexGrid defaults 2`] = `
Array [
  Object {
    "box-sizing": "border-box",
    "display": "flex",
    "flex-basis": "25%",
    "flex-direction": "column",
    "flex-grow": "1",
    "max-width": "100%",
    "position": "relative",
    "visibility": "visible",
  },
  Object {
    "box-sizing": "border-box",
    "display": "flex",
    "flex-basis": "25%",
    "flex-direction": "column",
    "flex-grow": "1",
    "max-width": "100%",
    "position": "relative",
    "visibility": "visible",
  },
  Object {
    "box-sizing": "border-box",
    "display": "flex",
    "flex-basis": "25%",
    "flex-direction": "column",
    "flex-grow": "1",
    "max-width": "100%",
    "position": "relative",
    "visibility": "visible",
  },
  Object {
    "box-sizing": "border-box",
    "display": "flex",
    "flex-basis": "25%",
    "flex-direction": "column",
    "flex-grow": "1",
    "max-width": "100%",
    "position": "relative",
    "visibility": "visible",
  },
  Object {
    "box-sizing": "border-box",
    "display": "flex",
    "flex-basis": "25%",
    "flex-direction": "column",
    "flex-grow": "1",
    "max-width": "100%",
    "position": "relative",
    "visibility": "visible",
  },
]
`;
